polynote-frontend/polynote/ui/component/notebook/common.ts

```diff
@@ -18,35 +19,69 @@ export function goToDefinition(
-        // TODO: if the location is a notebook cell?
-        if (result.source && result.location && result.location[0]) {
-            const loc = result.location[0];
-            const uri = Uri.parse(loc.uri);
-            const depSrc = {
-                language: uri.scheme,
-                content: result.source!,
-                position: { lineNumber: loc.line, column: loc.column },
-                sourceNotebook: notebookState
+
+        return result.location.map(loc => {
+            const absolute = loc.uri.startsWith("#") ? loc.uri : new URL(loc.uri, document.location.href).href;
+            return {
+                uri: Uri.parse(absolute),
+                range: posToRange(loc)
             }
-            ServerStateHandler.get.updateAsync(state => {
-                const dependencyLoaded = loc.uri in state.dependencySources;
-                const fileIsOpen = arrExists(state.openFiles, of => of.path === loc.uri);
-                return {
-                    dependencySources: updateProperty(
-                        loc.uri,
-                        dependencyLoaded ? setProperty("position", depSrc.position)
-                                         : setValue(depSrc)),
-                    openFiles: fileIsOpen ? NoUpdate
-                                          : append({ type: "dependency_source", path: loc.uri })
-                };
-            }).then(() => {
-                ServerStateHandler.selectFile(loc.uri)
-            })
-
-        }
-        return result.location.map(loc => ({
-            uri: Uri.parse(loc.uri),
-            range: posToRange(loc)
-        }))
+        })
     })
+}
+
+export function openDefinition(notebookState: NotebookStateHandler, lang: string, location: Location): Promise<void> {
+    const uriString = location.uri.toString(true);  // true means "don't render this URI incorrect by mangling its characters"
+
+    // even a relative "#X" URI gets turned into a file:///#X URI by the Microsoft thing. So really just have to
+    // assume that any fragment is a cell link
+    if (location.uri.fragment) {
+        const cellId = cellIdFromHash(location.uri.fragment);
+        notebookState.selectCellAt(cellId, Range.getStartPosition(location.range));
+        return Promise.resolve();
+    }
+
+    const params = new URLSearchParams(location.uri.query);
+    const filename = params.get("dependency");
+    const filePieces = filename?.split('.')
+    const fileLanguage = filePieces?.[filePieces.length - 1] || lang;
+
+    if (uriString in ServerStateHandler.state.dependencySources) {
+        return ServerStateHandler.get.updateAsync(state => {
+            const fileIsOpen = arrExists(state.openFiles, of => of.path === uriString);
+            return {
+                dependencySources: updateProperty(uriString, {position: Range.getStartPosition(location.range)}),
+                openFiles: fileIsOpen ? NoUpdate : append({type: "dependency_source", path: uriString})
```

Comment:
Hm, I wonder whether it would be possible to do the update to `openFiles` based on the update to `dependencySources`. Just to prevent needing to remember to update both?